File: python/src/loxtep/query.py
"""Query facade (MCP: loxtep_query)."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable

from .catalog import CatalogApi
from .discovery import DiscoveryApi


@dataclass
class QueryFacade:
    catalog: CatalogApi
    discovery: DiscoveryApi
    _query: Callable[..., Any]
    _list_tables: Callable[..., Any]
    _search: Callable[..., Any]

    def query(self, *args: Any, **kwargs: Any) -> Any:
        return self._query(*args, **kwargs)

    def list_tables(self, *args: Any, **kwargs: Any) -> Any:
        return self._list_tables(*args, **kwargs)

    def search(self, *args: Any, **kwargs: Any) -> Any:
        return self._search(*args, **kwargs)

File: nodejs/src/client/query.ts
/**
 * Query facade (MCP: loxtep_query).
 * Delegates to catalog, discovery, and data-product query/analytics methods.
 */

import type { createCatalogApi } from './catalog.js';
import type { createDiscoveryApi } from './discovery.js';
import type { createDataProductsApi } from './data-products.js';

export interface QueryFacadeDeps {
  catalog: ReturnType<typeof createCatalogApi>;
  discovery: ReturnType<typeof createDiscoveryApi>;
  data_products: ReturnType<typeof createDataProductsApi>;
}

export function createQueryFacade(deps: QueryFacadeDeps): {
  catalog: QueryFacadeDeps['catalog'];
  discovery: QueryFacadeDeps['discovery'];
  query: ReturnType<typeof createDataProductsApi>['query'];
  list_tables: ReturnType<typeof createDataProductsApi>['list_tables'];
  search: ReturnType<typeof createDataProductsApi>['search'];
} {
  return {
    catalog: deps.catalog,
    discovery: deps.discovery,
    query: deps.data_products.query.bind(deps.data_products),
    list_tables: deps.data_products.list_tables.bind(deps.data_products),
    search: deps.data_products.search.bind(deps.data_products),
  };
}

export type QueryFacade = ReturnType<typeof createQueryFacade>;
